// hono/dist/router/static-router/router.js
// src/router/static-router/router.ts
import { METHOD_NAME_ALL, METHODS, UnsupportedPathError } from "../../router.js";
var StaticRouter = class {
  constructor() {
    this.middleware = {};
    this.routes = {};
    [METHOD_NAME_ALL, ...METHODS].forEach((method) => {
      this.routes[method.toUpperCase()] = {};
    });
  }
  newRoute() {
    const route = {};
    const routeAll = this.routes[METHOD_NAME_ALL];
    Object.keys(routeAll).forEach((path) => {
      route[path] = {
        handlers: [...routeAll[path].handlers],
        params: {}
      };
    });
    return route;
  }
  add(method, path, handler) {
    var _a, _b;
    const { middleware, routes } = this;
    routes[method] || (routes[method] = this.newRoute());
    if (path === "/*") {
      path = "*";
    }
    if (path === "*") {
      if (method === METHOD_NAME_ALL) {
        middleware[_a = METHOD_NAME_ALL] || (middleware[_a] = { handlers: [], params: {} });
        Object.keys(middleware).forEach((m) => {
          middleware[m].handlers.push(handler);
        });
        Object.keys(routes).forEach((m) => {
          Object.values(routes[m]).forEach((matchRes) => matchRes.handlers.push(handler));
        });
      } else {
        middleware[method] || (middleware[method] = {
          handlers: [...middleware[METHOD_NAME_ALL]?.handlers || []],
          params: {}
        });
        middleware[method].handlers.push(handler);
        if (routes[method]) {
          Object.values(routes[method]).forEach((matchRes) => matchRes.handlers.push(handler));
        }
      }
      return;
    }
    if (/\*|\/:/.test(path)) {
      throw new UnsupportedPathError(path);
    }
    (_b = routes[method])[path] || (_b[path] = {
      handlers: [
        ...routes[METHOD_NAME_ALL][path]?.handlers || middleware[method]?.handlers || middleware[METHOD_NAME_ALL]?.handlers || []
      ],
      params: {}
    });
    if (method === METHOD_NAME_ALL) {
      Object.keys(routes).forEach((m) => {
        routes[m][path]?.handlers?.push(handler);
      });
    } else {
      routes[method][path].handlers.push(handler);
    }
  }
  match(method, path) {
    const { routes, middleware } = this;
    this.match = (method2, path2) => routes[method2][path2] || routes[METHOD_NAME_ALL][path2] || middleware[method2] || middleware[METHOD_NAME_ALL] || null;
    return this.match(method, path);
  }
};
export {
  StaticRouter
};
